observer/baseNodeProvider.go

```diff
@@ -25,8 +26,37 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
 
 	bop.mutNodes.Lock()
 	bop.nodes = newNodes
-	bop.allNodes = nodes
+	bop.allNodes = initAllNodesSlice(newNodes)
 	bop.mutNodes.Unlock()
 
 	return nil
 }
+
+func initAllNodesSlice(nodesOnShards map[uint32][]*data.NodeData) []*data.NodeData {
+	sliceToReturn := make([]*data.NodeData, 0)
+	shardIDs := make([]uint32, 0)
+	for shardID := range nodesOnShards {
+		shardIDs = append(shardIDs, shardID)
+	}
+	sort.SliceStable(shardIDs, func(i, j int) bool {
+		return shardIDs[i] < shardIDs[j]
+	})
+
+	finishedShards := make(map[uint32]struct{})
+	for i := 0; ; i++ {
```

Comment:
maybe break this code into smaller functions?

Reply:
extracted a function for returning the sorted shard ids